### PR TITLE
Migrating the toolkit to a client

### PR DESCRIPTION
**Pull request by fadeev**
Example:

```ts
import { ZetaChainClient } from "@zetachain/toolkit";

const client = new ZetaChainClient({network: "testnet"});
await client.getPools()
```

Endpoints can be customized:

```ts
const client = new ZetaChainClient({
  chains: {
    zeta_testnet: {
      api: [
        {
          type: "evm",
          url: "https://rpc.ankr.com/zetachain_evm_athens_testnet",
        },
      ],
    },
  },
  network: "testnet",
});
```

The design is very simple, `ZetaChainClient` is just a class for storing state (right now, that's `network` and `chains`). Methods are using the following `this` pattern to a) have access to class variables through `this` b) be defined in separate files for readability.

```ts
export const getEndpoints = function (
  this: ZetaChainClient, // <-- this
  type: any,
  network: any
) {
  if (!(this.chains as any)[network]) {
    throw new Error(`Network ${network} does not exist.`);
  }

  return (this.chains as any)[network].api.filter(
    (api: any) => api.type === type
  )[0]?.url;
};
```

## SDK

```ts
import { ZetaChainClient } from "@zetachain/toolkit/client";
import { ethers } from "ethers";

const client = new ZetaChainClient({
  network: "testnet",
  wallet: ethers.Wallet.fromMnemonic(process.env.MNEMONIC as string),
});

/**
 * Initiates a deposit transaction of native gas or ERC-20 assets as ZRC-20 from
 * a connected chain to ZetaChain.
 *
 * @param this - ZetaChainClient instance.
 * @param options - Deposit options.
 * @param options.chain - Label of the connected chain from which the deposit is
 * made.
 * @param options.amount - Amount to be deposited in human readable form.
 * @param options.erc20 - If an ERC-20 token is being deposited, the address of
 * the ERC-20 token contract. If not provided, the deposit is assumed to be in
 * native gas token.
 * @param options.message - If a message is specified, ZetaChain will deposit
 * tokens into the `recipient` contract and call with with the message as an argument.
 * @param options.recipient - Recipient address for the deposit. If not provided,
 * the deposit is made to the signer's address. If the message is provided, the
 * recipient is assumed to be a contract address.
 *
 * @returns A promise that resolves with the transaction details upon success.
 */
const { hash } = client.deposit({
  chain: "goerli_testnet",
  amount: "10000",
});

client.trackCCTX({ hash })

// Deposit ERC-20 to ZetaChain as ZRC-20 and call an omnichain contract (recipient) with a message
client.deposit({
  chain: "goerli_testnet",
  amount: "10000",
  erc20: "0x07865c6e87b9f70255377e024ace6630c1eaa37f", // if erc20 is provided, send erc20, otherwise: native gas
  recipient: "0x2cD3D070aE1BD365909dD859d29F387AA96911e1", // if recipient is not provided, deposit to sender
  message: [["string"], ["hello"]] // if a message is provided it's a "deposit and call"
}

/**
 * Initiates a withdraw transaction of a ZRC-20 token from ZetaChain to a
 * connected chain as a native gas or ERC-20 token.
 *
 * @param this - ZetaChainClient instance.
 * @param options - Withdrawal options.
 * @param options.amount - Amount to be withdrawn in human readable form.
 * @param options.zrc20 - ZRC-20 token contract address.
 * @param options.recipient - Recipient address for the withdrawal. If not provided,
 * the withdrawal is made to the signer's address.
 *
 * @returns A promise that resolves with the transaction details upon success.
 */
const tx = await client.withdraw({
  amount: "1",
  zrc20: "0x0cbe0dF132a6c6B4a2974Fa1b7Fb953CF0Cc798a",
});

/**
 *
 * Initiates a cross-chain transfer of ZETA tokens from the source chain to the
 * destination chain.
 *
 * @param this - ZetaChainClient instance.
 * @param options - Send ZETA options.
 * @param options.chain - Source chain label.
 * @param options.destination - Destination chain label.
 * @param options.amount - Amount of ZETA tokens to be sent in human readable form.
 * @param options.recipient - Optional recipient address for the token transfer. If not
 * provided, the token transfer is made to the signer's address.
 * @param options.gasLimit - Optional gas limit on the destination chain.
 *
 * @returns A promise that resolves with the transaction details upon success.
 */
const tx = await client.sendZeta({
  chain: "goerli_testnet",
  destination: "zeta_testnet",
  recipient: "0x4955a3F38ff86ae92A914445099caa8eA2B9bA32",
  amount: "3",
});

client.getBalances({
  evm: "0x2cD3D070aE1BD365909dD859d29F387AA96911e1"
})
```

I wanted to add an ability to send cross-chain messages as a method, but since the best practice is for contracts to have approved counterparty addresses (interactors), this means virtually no contract will be able to accept a message sent from a EOA.

### CODE REVIEW

**Comment by teguhsy** [comment]
Smooth

File: packages/client/src/client.ts
Comment (hernan-clich) [inline packages/client/src/client.ts:24]: Can `signer` and `wallet` really be anything?

File: packages/tasks/src/withdraw.ts
Comment (hernan-clich) [inline packages/tasks/src/withdraw.ts:18]: Let's remove

File: packages/tasks/src/sendZETA.ts
Comment (hernan-clich) [inline packages/tasks/src/sendZETA.ts:7]: What's the meaning of this? Can we remove it?

File: packages/tasks/src/faucet.ts
Comment (hernan-clich) [inline packages/tasks/src/faucet.ts:1]: What is this used for?

File: packages/tasks/src/deposit.ts
Comment (hernan-clich) [inline packages/tasks/src/deposit.ts:19]: Can we rename to something less ambiguous?

File: packages/tasks/src/deposit.ts
Comment (hernan-clich) [inline packages/tasks/src/deposit.ts:69]: Let's please remove

File: packages/tasks/src/balances.ts
Comment (hernan-clich) [inline packages/tasks/src/balances.ts:81]: Can we go back to `evmAddress` and `btcAddress`?

**Review by hernan-clich — APPROVED** [review]

**Review by hernan-clich — APPROVED** [review]
Approving as there isn't anything that prevents this from being merged and the main idea is to unblock merging, but I would strongly advise taking a look at the attached comments.

File: index.ts
Comment (fadeev) [reply index.ts]: This was just a file to test functionality, removed 90fa9a2

File: index.ts
Comment (fadeev) [reply index.ts]: Removed the whole file in https://github.com/zeta-chain/toolkit/commit/90fa9a226f90bd163c6f6f63eabcc1415ea8b909

File: index.ts
Comment (fadeev) [reply index.ts]: https://github.com/zeta-chain/toolkit/commit/90fa9a226f90bd163c6f6f63eabcc1415ea8b909

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts]: 55b805c

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts]: I've switched to using a helper to get ZetaChain's chain ID from the client to avoid hard-coding IDs.

```
chain_id: this.getChainId(`zeta_${this.network}`)
```

3a61455

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts:32]: I don't think we can as `this` is a keyword:

https://www.typescriptlang.org/docs/handbook/2/classes.html#this-parameters

Comment (fadeev) [reply packages/client/src/getBalances.ts:32]: The reason I'm using this pattern (sorry for the pun) is to define methods is separate files and still have access to the class instance through `this`.

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts]: Ok, 1c75e0d

I was kind of hoping to keep param names less verbose as this is an object, so users would have to repeat them.

```ts
getBalances({evm: "0x4955a3F38ff86ae92A914445099caa8eA2B9bA32"})

// vs

getBalances({evmAddress: "0x4955a3F38ff86ae92A914445099caa8eA2B9bA32"})
```

Do you think using `address` type instead of string would suffice or do you think being verbose is better.

```ts
// For example:
export const getBalances = async function (
  this: ZetaChainClient,
  { evm, btc }: { evm: address, btc?: address }
)
```

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts]: f7a2989

File: packages/client/src/deposit.ts
Comment (fadeev) [reply packages/client/src/deposit.ts:81]: faa7ac2

File: packages/client/src/deposit.ts
Comment (fadeev) [reply packages/client/src/deposit.ts:91]: cb35c5d

File: packages/client/src/getChainId.ts
Comment (fadeev) [reply packages/client/src/getChainId.ts:22]: Can't do it right now, because the exported type in the networks package does not match exactly what's in the networks JSON (specifically, it doesn't have chain aliases), created an issue for this: https://github.com/zeta-chain/networks/issues/44

File: packages/client/src/getFees.ts
Comment (fadeev) [reply packages/client/src/getFees.ts]: 1fd3156

File: packages/client/src/getFees.ts
Comment (fadeev) [reply packages/client/src/getFees.ts:30]: I've refactored `getBalances` so that it no longer has chain ID hard-coded. e0efa94

File: packages/client/src/getFees.ts
Comment (fadeev) [reply packages/client/src/getFees.ts]: 0bdc652

File: packages/client/src/getForeignCoins.ts
Comment (fadeev) [reply packages/client/src/getForeignCoins.ts:8]: I agree, we need to fix the types we generate and publish from the node. Right now they're unusable, because they don't match the format returned by the API: https://github.com/zeta-chain/node/issues/1795

File: packages/client/src/getPools.ts
Comment (fadeev) [reply packages/client/src/getPools.ts]: The problem was with generated types for protocol contracts, I've released a new RC version, now it's good.

f339a34

File: packages/client/src/getBalances.ts
Comment (fadeev) [reply packages/client/src/getBalances.ts]: It should not be undefined, so I've added a check: a815664

File: packages/client/src/getPools.ts
Comment (fadeev) [reply packages/client/src/getPools.ts]: It should not be undefined, so I've added a check: a815664

File: packages/client/src/getSupportedChains.ts
Comment (fadeev) [reply packages/client/src/getSupportedChains.ts:8]: I would really prefer not to hard-code them.

File: packages/client/src/sendZeta.ts
Comment (fadeev) [reply packages/client/src/sendZeta.ts]: 27b14d6

File: packages/client/src/sendZeta.ts
Comment (fadeev) [reply packages/client/src/sendZeta.ts:60]: 15ee993

Comment (fadeev) [reply packages/client/src/sendZeta.ts]: 15ee993

File: index.ts
Comment (hernan-clich) [reply index.ts]: Why was I even able to see this in the first place then? 🫣
I guess their integrated vs code editor (the one that opens when pressing the "." button in the UI) might be buggy and contained and older state.

File: packages/client/src/getBalances.ts
Comment (hernan-clich) [reply packages/client/src/getBalances.ts:32]: Feel free to leave as is then! 👌

File: packages/client/src/client.ts
Comment (fadeev) [reply packages/client/src/client.ts:24]: 1ef37fa

File: packages/tasks/src/withdraw.ts
Comment (fadeev) [reply packages/tasks/src/withdraw.ts:18]: 65a54ec

File: packages/tasks/src/sendZETA.ts
Comment (fadeev) [reply packages/tasks/src/sendZETA.ts:7]: 720e529

File: packages/tasks/src/faucet.ts
Comment (fadeev) [reply packages/tasks/src/faucet.ts:1]: The faucet doesn't ship with types, this tells TS code to not look for types for this package.

File: packages/tasks/src/deposit.ts
Comment (fadeev) [reply packages/tasks/src/deposit.ts:19]: a35745f

File: packages/tasks/src/deposit.ts
Comment (fadeev) [reply packages/tasks/src/deposit.ts]: https://github.com/zeta-chain/toolkit/commit/65a54eca3b87da4e09255bd228973932a48e4df5

File: packages/tasks/src/balances.ts
Comment (fadeev) [reply packages/tasks/src/balances.ts]: https://github.com/zeta-chain/toolkit/commit/1c75e0d8148eb8b0c06844cf293b64016bc570e8

File: packages/client/src/getBalances.ts
Comment (hernan-clich) [reply packages/client/src/getBalances.ts]: I would still add the "address" suffix even if it's more verbose, but totally optional.

**Comment by fadeev** [comment]
@lucas-janon @andresaiello can you, please, review? I need codeowners review before merging. Thanks!

**Review by andresaiello — APPROVED** [review]

**Pull request merged**
